File: src/main/java/com/github/alexandrecarlton/idea/settings/layout/tools/file_watchers/FileWatcherSettings.java
package com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import org.immutables.value.Value;

import java.util.Optional;

@Value.Immutable
@JsonDeserialize(as = ImmutableFileWatcherSettings.class)
public interface FileWatcherSettings {

  String name();

  FileWatcherFilesToWatchSettings filesToWatch();

  FileWatcherToolToRunOnChangesSettings toolToRunOnChanges();

  Optional<FileWatcherAdvancedOptionsSettings> advancedOptions();
}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/layout/tools/file_watchers/FileWatcherSettings.kt
package com.github.alexandrecarlton.idea.settings.layout.tools.file_watchers

import com.fasterxml.jackson.annotation.JsonProperty

data class FileWatcherSettings(

    @JsonProperty("Name")
    val name: String,

    @JsonProperty("Files to watch")
    val filesToWatch: FileWatcherFilesToWatchSettings,

    @JsonProperty("Tool to Run on Changes")
    val toolToRunOnChanges: FileWatcherToolToRunOnChangesSettings,

    @JsonProperty("Advanced Options")
    val advancedOptions: FileWatcherAdvancedOptionsSettings? = null
)
